LobbyTestUpload ревью реквест

Aher0nt opened the pull request:
Upload Lobby Branch + Test lobby scene with test code

Figer93 committed:
LobbyTestUpload

Upload Lobby Branch + Test lobby scene with test code

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,53 @@
+
+public class TestLobby : MonoBehaviour
+{
+    public TMP_InputField inputLobbyCode;
```
публичные поля пишутся с большой буквы

Aher0nt replied:
лучше вообще через SerializeField писать и в инспектор кидать

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,53 @@
+
+    private async void Start()
+    {
+        await UnityServices.InitializeAsync();
```
готова изоляция для этого метода в интерфейсе IUnityService

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,53 @@
+            _lobby = await LobbyService.Instance.CreateLobbyAsync(lobbyName, maxPlayers);
+            Debug.Log(_lobby.Name + $" lobby created with {_lobby.MaxPlayers} max players. Lobby code: " + _lobby.LobbyCode);
+        }
+        catch (LobbyServiceException exception)
```
позволительно вместо exception писать e

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,53 @@
+            string lobbyName = "MyLobby";
+            int maxPlayers = 4;
+
+            _lobby = await LobbyService.Instance.CreateLobbyAsync(lobbyName, maxPlayers);
```
Надо огородить LobbyService интерфейсом

Figer93 committed:
Update of Lobby code

refactoring + adding new lobby features

Figer93 committed:
Revert "Update of Lobby code"

This reverts commit 50114726351394d9df6d9e9e02586348b37e600d.

Aher0nt committed:
Revert "Revert "Update of Lobby code""

This reverts commit fcef5a38617c063e10e5fcbf9ff130b75f352308.

Aher0nt commented on Assets/Scripts/Lobby/ILobbyProvider.cs:
```diff
@@ -0,0 +1,6 @@
+public interface ILobbyProvider
+{
+    public void ConnectToLobby(string lobbyCode);
+    public void GetLobbies(string lobbyName, int maxPlayers);
```
весь функционал повторен в ILobbyService. Удалить

Aher0nt commented on Assets/Scripts/Lobby/ILobbyService.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public void CreateLobby(string lobbyName, int maxPlayers);
+    public void JoinLobbyByCode(string lobbyCode);
+    public void GetLobbiesList();
```
Get не может возвращать void
множественное окончание уже подразумевает, что ты вернешь List

Aher0nt commented on Assets/Scripts/Lobby/ILobbyService.cs:
```diff
@@ -0,0 +1,9 @@
+    public void CreateLobby(string lobbyName, int maxPlayers);
+    public void JoinLobbyByCode(string lobbyCode);
+    public void GetLobbiesList();
+    public void PrintPlayers(Lobby lobby);
```
перекидываешь ответственность вывода на лобби. Нарушение SRP

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+public class TestLobby : MonoBehaviour, ILobbyService
+{
+    public TMP_InputField InputLobbyCode;
+    public TextMeshProUGUI LobbyCode;
```
Этот класс реализует лобби. Он не может хранить ссылки на UI. Нарушение SRP

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+    
+    private Lobby _hostLobby;
+    private float _heartbeatTimer;
+    private string _playerName;
```
_playerName потом должно вытаскиваться из БД

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+    private void Awake()
+    {
+        _playerName = "Player" + UnityEngine.Random.Range(0, 1000);
+        Debug.Log(_playerName + " successfully logged in"); // test
```
никуда он не LoggedIn, ты аутентификацию нигде не вызвал

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+
+    private void Update()
+    {
+        HandleLobbyHeartbeat();
```
Для этого есть интерфейс ITickable от Zenject

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+            _heartbeatTimer -= Time.deltaTime;
+            if (_heartbeatTimer <= 0f)
+            {
+                float heartbeatTimerMax = 15;
```
Это должно хранится в константах в шапке класса

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+                float heartbeatTimerMax = 15;
+                _heartbeatTimer = heartbeatTimerMax;
+
+                await LobbyService.Instance.SendHeartbeatPingAsync(_hostLobby.Id);
```
ты не инициализировал сервисы юнити в public void Init()

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+                {
+                    Data = new Dictionary<string, PlayerDataObject>
+                    {
+                        {"PlayerName", new PlayerDataObject(PlayerDataObject.VisibilityOptions.Member, _playerName)}
```
уффф, шестой уровень вложенности. Чисто в теории, это очень плохо

Aher0nt replied:
лучше тогда отдельно написать PlayerBuilder по паттерну Builder

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+            Lobby lobby = await LobbyService.Instance.CreateLobbyAsync(lobbyName, maxPlayers, createLobbyOptions);
+            _hostLobby = lobby;
+            Debug.Log(lobby.Name + $" lobby created with {lobby.MaxPlayers} max players. Lobby code: " + lobby.LobbyCode);
+            LobbyCode.text = lobby.LobbyCode;
```
Нарушение SRP

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+
+    public async void JoinLobbyByCode(string lobbyCode)
+    {
+        lobbyCode = InputLobbyCode.text;
```
есть моменты, когда lobbyCode == null, поэтому лучше хранить его только внутри метода

Aher0nt replied:
а, я слепой, не заметил, что это параметр

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,131 @@
+    {
+        try
+        {
+            QueryLobbiesOptions queryLobbiesOptions = new QueryLobbiesOptions
```
сюда тоже нужен отдельный класс для конфигурации запроса, т.к. нарушение SRP

Aher0nt commented on Assets/Scripts/UI/Lobby/CreateLobbyButton.cs:
```diff
@@ -0,0 +1,30 @@
+{
+    [SerializeField] private Button _createNewLobbyButton;
+    [SerializeField] private TMP_InputField _lobbyName;
+    [SerializeField] private TMP_InputField _maxPlayers;
```
когда пишем переменные, которые содержат UI, всегда придерживаемся венгерской нотации, как в строке 8.
Почитай правила именования в #объявления дискорда

Figer93 replied:
Типа так btnCreateNewLobby?
Слышал что эти наименования уже не актуальны, так как у языка уже есть сильная система типов и использование префиксов не требуется, кто знает..

Hitsar replied:
private Button _createNewLobby**Button**;
private TMP_InputField _lobbyName**InputField**;
так пишешь

Aher0nt commented on Assets/Scripts/UI/Lobby/CreateLobbyButton.cs:
```diff
@@ -0,0 +1,30 @@
+
+    private void CreateNewLobby() //Not working
+    {
+        int maxPlayers = int.Parse(_maxPlayers.text);
```
мне кажется, тут можно просто explicit cast делать через (int). Поправь, если нет

Hitsar replied:
Не работает так

Figer93 committed:
Upload for Hitsar

Hitsar committed:
Сцен контекст для лобби инстайлера, добавил сцену лобби+аутификация

Figer93 committed:
Update testcode for Hitsar

Figer93 committed:
Upload for tests

Hitsar committed:
вывод игроков при заходе

Hitsar committed:
фиговый солид для TestLobby

Hitsar committed:
фиговый солид для UI лобби

надо реализовать несколько механик, и поправить мой фиговый нейминг

Hitsar committed:
перенёс методы Start в Construct (исправил ошибки)

Hitsar committed:
Update LobbyOwner.cs

player.Id

Aher0nt commented on Assets/Scripts/AppInstaller.cs:
```diff
@@ -1,5 +1,6 @@
 using Assets.Scripts.Authentication.DI;
 using Assets.Scripts.UnityService.DI;
+using Lobby.DI;
```
придерживайтесь общего наименования неймспейсов

Hitsar replied:
зачем лишний раз писать Assets.Scripts если мы и так в папке Scripts

Aher0nt commented on Assets/Scripts/Lobby/Interfaces/ILobbyOwner.cs:
```diff
@@ -0,0 +1,13 @@
+        void CreateLobby(string lobbyName, int maxPlayers, CreateLobbyOptions createLobbyOptions);
+        void DeleteLobby();
+        void RemovePlayer(Player player);
+        Unity.Services.Lobbies.Models.Lobby GetHostLobby();
```
смысл это интерфейса в том, чтобы изолировать зависимости, а не тянуть их дальше в проект

Aher0nt commented on Assets/Scripts/Lobby/Interfaces/ILobbySystem.cs:
```diff
@@ -0,0 +1,12 @@
+    public interface ILobbySystem
+    {
+        void HandleLobbyHeartbeat();
+        void PrintPlayers(Unity.Services.Lobbies.Models.Lobby lobby);
```
нарушение SRP

Aher0nt commented on Assets/Scripts/Lobby/Interfaces/ILobbySystem.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface ILobbySystem
+    {
+        void HandleLobbyHeartbeat();
```
зачем внешним классам знать про Heartbeat

Aher0nt commented on Assets/Scripts/Lobby/Interfaces/ILobbySystem.cs:
```diff
@@ -0,0 +1,12 @@
+        void HandleLobbyHeartbeat();
+        void PrintPlayers(Unity.Services.Lobbies.Models.Lobby lobby);
+        void LobbiesList(QueryLobbiesOptions queryLobbiesOptions);
+        void RefreshLobbiesList();
```
Нарушение SRP

Aher0nt commented on Assets/Scripts/Lobby/LobbyMember.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    public class LobbyMember : ILobbyMember
+    {
+        private Unity.Services.Lobbies.Models.Lobby _lobby;
```
юзингами пользуйся

Aher0nt commented on Assets/Scripts/Lobby/LobbyMember.cs:
```diff
@@ -0,0 +1,45 @@
+                Unity.Services.Lobbies.Models.Lobby lobby = await Lobbies.Instance.JoinLobbyByCodeAsync(code);
+                _lobby = lobby;
+                Debug.Log("Joined lobby with name " + lobby.Name);
+                _lobbySystem.PrintPlayers(lobby);
```
в чем смысл разбрасывать логику лобби на два слоя

Aher0nt commented on Assets/Scripts/Lobby/LobbyMember.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            try
+            {
+                await Lobbies.Instance.RemovePlayerAsync(_lobby.Id, AuthenticationService.Instance.PlayerId);
```
нарушение зависимостей. Уже есть готовый интерфейс, изолирующий AuthenticationService. Теперь его больше НИГДЕ НЕЛЬЗЯ ИСПОЛЬЗОВАТЬ

Aher0nt commented on Assets/Scripts/Lobby/LobbySystem.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public async void HandleLobbyHeartbeat()
+        {
+            if (_hostLobby == null) return;
```
какая же она System, если контролирует только один лобби

Aher0nt commented on Assets/Scripts/Lobby/LobbySystem.cs:
```diff
@@ -0,0 +1,68 @@
+            _heartbeatTimer -= Time.fixedDeltaTime;
+            if (_heartbeatTimer <= 0f)
+            {
+                float heartbeatTimerMax = 15;
```
Вынести в константы, подписать размерность

Aher0nt commented on Assets/Scripts/Lobby/LobbySystem.cs:
```diff
@@ -0,0 +1,68 @@
+                float heartbeatTimerMax = 15;
+                _heartbeatTimer = heartbeatTimerMax;
+
+                await LobbyService.Instance.SendHeartbeatPingAsync(_hostLobby.Id);
```
ты не инициализировал сервисы юнити

Aher0nt commented on Assets/Scripts/Lobby/LobbySystem.cs:
```diff
@@ -0,0 +1,68 @@
+        {
+            try
+            {
+                QueryResponse queryResponse = await Lobbies.Instance.QueryLobbiesAsync(queryLobbiesOptions);
```
и? ты потом данные собираешься из дебага забирать?

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,127 @@
+*    public void Initialize()
+*    {
+*        _playerName = "Player" + Random.Range(0, 1000);
+*        Debug.Log(_playerName + " successfully logged in");
```
аутентификация не проверена

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,127 @@
+*        HandleLobbyHeartbeat();
+*    }
+*
+*    private async void HandleLobbyHeartbeat()
```
дублирование кода

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,127 @@
+*                {
+*                    Data = new Dictionary<string, PlayerDataObject>
+*                    {
+*                        {"PlayerName", new PlayerDataObject(PlayerDataObject.VisibilityOptions.Member, _playerName)}
```
для этой башни надо написать LobbyBuilder

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,127 @@
+*        }
+*    }
+*
+*    public async void JoinLobbyByCode(string lobbyCode)
```
дублирование

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,127 @@
+*        }
+*    }
+*
+*    public async void LobbiesList()
```
дублирование

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,127 @@
+*    {
+*        try
+*        {
+*            QueryLobbiesOptions queryLobbiesOptions = new QueryLobbiesOptions
```
надо написать билдер для этого запроса

Figer93 replied:
Будет сделано, сэр

Aher0nt commented on Assets/Scripts/Lobby/TestLobby.cs:
```diff
@@ -0,0 +1,127 @@
+*        }
+*    }
+*
+*    public void PrintPlayers(Lobby lobby)
```
Этого метода не должно быть в готовом коде

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyMemberUI.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        [SerializeField] private Button _joinLobbyButton;
+        [SerializeField] private Button _exitLobbyButton;
+        [SerializeField] private TMP_InputField _lobbyCode;
```
нарушил венгерскую нотацию

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyOwnerUI.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        [SerializeField] private Button _createNewLobbyButton;
+        [SerializeField] private Button _deleteLobbyButton;
+        [SerializeField] private TMP_InputField _lobbyName;
```
нарушил венгерскую нотацию

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbySystemUI.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace UI.Lobby
+{
+    public class LobbySystemUI : MonoBehaviour
```
это просто обвес на кнопку, а не UI системы лобби

Hitsar committed:
переместил логику в 1 скрипт, список игроков и лобби в UI монобехи (временно дебагом)

Aher0nt commented on Assets/Scripts/Lobby/Interfaces/ILobbyService.cs:
```diff
@@ -1,13 +1,13 @@
         void DeleteLobby();
-        void RemovePlayer(Player player);
+        void JoinLobbyByCode(string code);
+        void ExitLobby(string id);
```
ты можешь выйти только из того лобби, в котором ты находишься. Не забываем, что этот код выполняется на стороне клиента

Aher0nt replied:
Еще нужен метод, чтобы кикнуть игрока из лобби

Hitsar replied:
я сделал метод ExitLobby(string playerId) если игрок сам выходит то вставляется его айди, если его кикают то вставляется айди того кого кикают (это считывается в другом скрипте), как по мне нету смысла разделять, получится дубляж кода

Figer93 replied:
Added

Aher0nt commented on Assets/Scripts/Lobby/Interfaces/ILobbyService.cs:
```diff
@@ -1,13 +1,13 @@
-        void RemovePlayer(Player player);
+        void JoinLobbyByCode(string code);
+        void ExitLobby(string id);
         Unity.Services.Lobbies.Models.Lobby GetHostLobby();
```
используй using

Hitsar replied:
не получается

Aher0nt commented on Assets/Scripts/Lobby/LobbyService.cs:
```diff
@@ -0,0 +1,80 @@
+{
+    public class LobbyService : ILobbyService, IFixedTickable
+    {
+        private Unity.Services.Lobbies.Models.Lobby _hostLobby;
```
используй using

Hitsar replied:
не получается

Aher0nt commented on Assets/Scripts/Lobby/LobbyService.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            try
+            {
+                Unity.Services.Lobbies.Models.
```
используй using

Hitsar replied:
не получается

Aher0nt replied:
тогда переименуйте неймспейсы на LobbyService

Aher0nt commented on Assets/Scripts/Lobby/LobbyService.cs:
```diff
@@ -0,0 +1,80 @@
+            try
+            {
+                Unity.Services.Lobbies.Models.Lobby lobby = await Lobbies.Instance.JoinLobbyByCodeAsync(code);
+                _hostLobby = lobby;
```
так это тогда не _hostLobby, а CurrentLobby

Aher0nt replied:
норм будет, когда в лобби все 10 человек каждые 15 секунд будут слать пинги? Надо оптимизировать

Aher0nt commented on Assets/Scripts/Lobby/LobbyService.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            try
+            {
+                await Lobbies.Instance.RemovePlayerAsync(_hostLobby.Id, id);
```
NullPointerException

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbiesPrinter.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace UI.Lobby
+{
+    public class LobbiesPrinter : MonoBehaviour, ILobbiesPrinter
```
В UI не надо интерфейсы к объектам лепить, их все равно никто подтягивать не будет. Ты сам ими не пользуешься

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbiesPrinter.cs:
```diff
@@ -0,0 +1,42 @@
+using UI.Lobby.Interfaces;
+using Unity.Services.Lobbies;
+using Unity.Services.Lobbies.Models;
```
UI не должен знать про Unity.Services

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbiesPrinter.cs:
```diff
@@ -0,0 +1,42 @@
+        {
+            try
+            {
+                QueryResponse queryResponse = await Lobbies.Instance.QueryLobbiesAsync(_queryLobbiesOptions);
```
Эта логика должна быть в модуле, в UI весь код пишется только для отображения информации

Aher0nt replied:
NullPointerException, ты нигде не вызвал метод FilterOptions

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyPlayersPrinter.cs:
```diff
@@ -0,0 +1,30 @@
+    public class LobbyPlayersPrinter : MonoBehaviour, IPlayersPrinter
+    {
+        private ILobbyService _lobbyService;
+        private Unity.Services.Lobbies.Models.Lobby _hostLobby;
```
используй using

Hitsar replied:
не получается

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyPlayersPrinter.cs:
```diff
@@ -0,0 +1,30 @@
+        private void Init(ILobbyService lobbyService)
+        {
+            _lobbyService = lobbyService;
+            _hostLobby = _lobbyService.GetHostLobby();
```
Singleton Exception. Inject-методы исполняются до инициализации сервисов.

Aher0nt commented on Assets/Scripts/Lobby/LobbyService.cs:
```diff
@@ -0,0 +1,80 @@
+
+namespace Lobby
+{
+    public class LobbyService : ILobbyService, IFixedTickable
```
ты не подтянул зависимость от ILobbyService. Нельзя надеяться, что его кто-то поднимет вместо тебя

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyPlayersPrinter.cs:
```diff
@@ -0,0 +1,30 @@
+        public void PrintPlayers()
+        {
+            Debug.Log("Players in Lobby " + _hostLobby.Name);
+            foreach (Player player in _hostLobby.Players)
```
у нас должны быть своя модель Player, которую нам потом будет удобно в БД хранить. Этот Player можно использовать только для связи непосредственно с сервисом. Чтобы создать однозначность, нашего player можно хранить внутри playera юнити, т.к. это будет DTO (гугл)

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyServiceUI.cs:
```diff
@@ -0,0 +1,47 @@
+        
+        [SerializeField] private Button _joinLobbyButton;
+        [SerializeField] private Button _exitLobbyButton;
+        [SerializeField] private TMP_InputField _lobbyCode;
```
слишком комплексный UI, надо резделить на части

Figer93 committed:
Small Lobby code update

Figer93 committed:
Lobby code refactoring

Figer93 committed:
Additional refactoring

Hitsar committed:
небольшой рефактор

Hitsar committed:
Merge branch 'Lobby' of https://github.com/Moba-for-pc/Moba into Lobby

Figer93 committed:
Division of the complex class into subclasses

Hitsar committed:
Update LobbyService.cs

Hitsar committed:
Merge branch 'Lobby' of https://github.com/Moba-for-pc/Moba into Lobby

Figer93 committed:
Lobby code refactoring

Figer93 committed:
Merge branch 'Lobby' of https://github.com/Moba-for-pc/Moba into Lobby

Aher0nt commented on Assets/Scripts/Lobby/DI/LobbyInstaller.cs:
```diff
@@ -0,0 +1,15 @@
+        {
+            Container.BindInterfacesTo<LobbyService>().AsSingle();
+            Container.Bind<IPlayersPrinter>().To<LobbyPlayersDisplay>().AsSingle();
+            Container.Bind<ILobbiesPrinter>().To<LobbiesDisplay>().AsSingle();
```
зачем вам интерфейсы на UI

Aher0nt commented on Assets/Scripts/Lobby/ILobbyService.cs:
```diff
@@ -0,0 +1,13 @@
+        void CreateLobby(string lobbyName, int maxPlayers, CreateLobbyOptions createLobbyOptions);
+        void DeleteLobby();
+        void JoinLobbyByCode(string code);
+        void ExitLobby(string id);
```
Ты сам должен знать, в каком лобби сейчас находится твой игрок. Тыж тут главный

Figer93 replied:
ExitLobby(string id); ?

Aher0nt commented on Assets/Scripts/Lobby/LobbyService.cs:
```diff
@@ -0,0 +1,81 @@
+        {
+            try
+            {
+                await Lobbies.Instance.RemovePlayerAsync(_currentLobby.Id, id);
```
нигде нет проверок на null

Hitsar replied:
Этот метод сможет вызываться ТОЛЬКО из лобби, поэтому в любом случае null он не будет

Aher0nt replied:
а ты откуда знаешь, куда умные дизайнеры решат запихнуть эту кнопку? Код не должен падать из-за того, что ему снаружи что-то не так сделали. Нам надо построить такую стену, чтобы она не упала даже при максимально неправильном использовании

Aher0nt commented on Assets/Scripts/UI/Lobby/ExitLobbyButton.cs:
```diff
@@ -0,0 +1,25 @@
+            _exitLobbyButton.onClick.AddListener(ExitLobby);
+        }
+        
+        private void ExitLobby() => _lobbyService.ExitLobby(AuthenticationService.Instance.PlayerId);
```
вы не можете пользоваться этим сервисом

Figer93 replied:
Тогда как быть

Aher0nt commented on Assets/Scripts/UI/Lobby/ILobbiesPrinter.cs:
```diff
@@ -0,0 +1,8 @@
+    public interface ILobbiesPrinter
+    {
+        void PrintLobbies();
+        void FilterOptions();
```
Зачем он вам?

Hitsar replied:
что бы вызывать методы по кнопкам

Aher0nt commented on Assets/Scripts/UI/Lobby/IPlayersPrinter.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface IPlayersPrinter
+    {
+        void PrintPlayers();
```
Кому могут понадобиться эти интерфейсы?

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,54 @@
+        {
+            try
+            {
+                QueryResponse queryResponse = await Lobbies.Instance.QueryLobbiesAsync(_queryLobbiesOptions);
```
В UI не должно быть логики

Hitsar replied:
Это не логика, мы применяем фильтр и далее результаты с ним выводим
Фильтр пишет нижу, это тоже логикой назвать нельзя, мы кнопками устанавливаем значения в нём

Aher0nt replied:
это логика, потому что ты этот запрос хранишь в переменных класса(_queryLobbiesOptions) + используешь сервисы юнити вмето обертки. Значит, вы свою работу по изоляции сделали плохо, раз она не покрывает все потребности интерфейса

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,54 @@
+            }
+        }
+
+        public void FilterOptions()
```
не вызван

Figer93 replied:
нету применения

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    public class LobbiesDisplay : MonoBehaviour, ILobbiesPrinter
+    {
+        [SerializeField] private Button _refreshButton;
```
рефреш не должен ничего выводить, он просто отправляет запрос на обновление

Hitsar replied:
Обновление по сути и есть повторный вывод но с новыми данными

Aher0nt replied:
у тебя потом вывод будут делать другие элементы интерфейса, у которых не будет доступа к RefreshButton. Им надо какое-нибудь событие о том, что лобби обновились

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyPlayersDisplay.cs:
```diff
@@ -0,0 +1,30 @@
+        public void PrintPlayers()
+        {
+            Unity.Services.Lobbies.Models.
+                Lobby currentLobby = _lobbyService.GetHostLobby();
```
Здесь надо пользоваться своим классов лобби/ своим списком игроков, а не тащить классы у юнити сервисов

Aher0nt replied:
неприятно, когда со стороны выходит обновление, и у тебя может упасть не только один модуль, а весь UI, который с ним связан

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyServiceUI.cs:
```diff
@@ -0,0 +1,61 @@
+/*
+using TMPro;
+using Unity.Services.Authentication;
+using Unity.Services.Lobbies;
```
эт что за артхаус

Aher0nt replied:
мы НЕ ХРАНИМ код в комментариях. Если он нам сильно понадобится, его можно взять из истории git

Figer93 committed:
Added builder for query lobby and lobby options.

Aher0nt commented on Assets/Scripts/Lobby/Builder/CreateLobbyOptionsBuilder.cs:
```diff
@@ -0,0 +1,43 @@
+using System.Collections.Generic;
+using Unity.Services.Lobbies.Models;
+
+namespace Unity.Services.Lobbies.Builder
```
ну у тебя ж тут не один билдер, а два

Aher0nt commented on Assets/Scripts/Lobby/Builder/QueryLobbiesOptionsBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+            return this;
+        }
+
+        public QueryLobbiesOptionsBuilder SetSkip(int skip)
```
не очень очевидное название метода. Раз это наш код, необязательно копировать в имя метода название изменяемого свойства

Figer93 replied:
SetSkipValue

Aher0nt commented on Assets/Scripts/Lobby/Builder/QueryLobbiesOptionsBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+            return this;
+        }
+
+        public QueryLobbiesOptionsBuilder SetContinuationToken(string continuationToken)
```
имя метода не обязательно должно начинаться с Set, можно так же использовать префиксы Make, Add, Use, etc.

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbiesDisplay.cs:
```diff
@@ -36,19 +36,15 @@ public async void PrintLobbies()
-                }
-            };
+            _queryLobbiesOptions = new QueryLobbiesOptionsBuilder()
+                .Reset()
```
ты создаешь новый билдер, в нем точно пустой запрос. Это типа "гарантируется" паттерном билдера, что при его создании у запроса дефолтные настройки

Aher0nt commented on Assets/Scripts/UI/Lobby/LobbyCreationButton.cs:
```diff
@@ -26,8 +27,16 @@ private void Init(ILobbyService lobbyService)
             int maxPlayers = int.Parse(_maxPlayersInputField.text);
+            
+            CreateLobbyOptions lobbyOptions = new CreateLobbyOptionsBuilder()
+                .Reset()
```
тут тоже можно убрать

Figer93 committed:
Division into classes, refactoring

Aher0nt commented on Assets/Scripts/Lobby/Builders/QueryLobbiesOptionsBuilderImp.cs:
```diff
@@ -0,0 +1,21 @@
+    public QueryLobbiesOptions BuildLobbyQueryOptions()
+    {
+        QueryLobbiesOptions lobbyOptions = new QueryLobbiesOptionsBuilder()
+            .Reset()
```
бесполезная операция

Aher0nt commented on Assets/Scripts/Lobby/Builders/QueryLobbiesOptionsBuilderImp.cs:
```diff
@@ -0,0 +1,21 @@
+using Unity.Services.Lobbies.Builders;
+using UnityEngine;
+
+public class QueryLobbiesOptionsBuilderImp : MonoBehaviour
```
что за грязь
непонятно кого имплементирует, да еще и от монобеха наследуется

Aher0nt commented on Assets/Scripts/Lobby/DI/LobbyInstaller.cs:
```diff
@@ -0,0 +1,15 @@
+        {
+            Container.BindInterfacesTo<LobbyService>().AsSingle();
+            Container.Bind<IPlayersPrinter>().To<LobbyPlayersDisplay>().AsSingle();
+            Container.Bind<ILobbiesDisplay>().To<LobbiesDisplay>().AsSingle();
```
Ну вот вы мне просто объясните, зачем вам нужны эти принтеры и дисплеи?

Hitsar committed:
Добавил DTO Player & Lobby. рефакторинг

в аутификации добавить присвоение данных плееру

Aher0nt commented on Assets/Scripts/DTO/DI/DtoInstaller.cs:
```diff
@@ -0,0 +1,13 @@
+        public override void InstallBindings()
+        {
+            Container.Bind<Player>().AsSingle();
+            Container.Bind<Lobby>().AsSingle();
```
Это модель данных, ее создают сами объекты чтобы обмениваться друг с другом. Ей не нужен DI

Aher0nt commented on Assets/Scripts/AppInstaller.cs:
```diff
@@ -11,6 +13,8 @@ public override void InstallBindings()
             AuthenticationInstaller.Install(Container);
             UnityServiceInstaller.Install(Container);
+            LobbyInstaller.Install(Container);
+            DtoInstaller.Install(Container);
```
моделям не нужен инсталллер

Aher0nt commented on Assets/Scripts/DTO/Lobby.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public class Lobby
+    {
+        public Unity.Services.Lobbies.Models.Lobby CurrentLobby;
```
НЕТ!! Ты создавал модель, чтобы уйти от юнити сервисов, а не сделать их частью себя!

Aher0nt commented on Assets/Scripts/DTO/Player.cs:
```diff
@@ -0,0 +1,10 @@
+        public string Name;
+        public string Id;
+        public Lobby CurrentLobby;
+        public Unity.Services.Lobbies.Models.Player PlayerModel;
```
Аналогично

Aher0nt commented on Assets/Scripts/LobbyService/Builders/LobbyOptionsBuilder.cs:
```diff
@@ -0,0 +1,43 @@
+        public LobbyOptionsBuilder Reset()
+        {
+            _lobbyOptions = new CreateLobbyOptions();
+            return this;
```
создаешь LobbyOption в двух местах. Лучше тогда reset сразу в конструкторе вызвать

Aher0nt commented on Assets/Scripts/LobbyService/Builders/QueryLobbiesOptionsBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+        public QueryLobbiesOptionsBuilder Reset()
+        {
+            _queryLobbiesOptions = new QueryLobbiesOptions();
+            return this;
```
то же самое

Aher0nt commented on Assets/Scripts/LobbyService/DI/LobbyInstaller.cs:
```diff
@@ -0,0 +1,14 @@
+        public override void InstallBindings()
+        {
+            Container.BindInterfacesTo<LobbyService>().AsSingle();
+            Container.Bind<LobbyOptionsBuilder>().AsSingle();
```
а второй билдер не пригодился?
Билдеры добавляются как Transient, потому что нужны для выполнения единицы работы

Aher0nt commented on Assets/Scripts/LobbyService/ILobbyService.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public interface ILobbyService
+    {
+        void CreateLobby(string lobbyName = "New lobby", int maxPlayers = 6, bool isPrivate = false, Player player = null, Dictionary<string, DataObject> data = null);
```
в интерфейсе параметры по умолчанию не пишем, это затрудняет задачу юнит-тестирования

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+using Unity.Services.Lobbies.Models;
+using UnityEngine;
+using Zenject;
+using Lobby = Assets.Scripts.DTO.Lobby;
```
DTO не выделяют в отдельную папку, это просто паттерн

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+    {
+        private LobbyOptionsBuilder _lobbyOptionsBuilder;
+        private Lobby _lobby;
+        private Player _player;
```
А других Player тебе разве не надо знать?

Hitsar replied:
у каждого пользователя свой плеер и лобби сервис, других я буду знать из _lobby

Aher0nt replied:
а зачем себя хранить, ты себя тоже можешь из лобби взять

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            _lobbyOptionsBuilder = lobbyOptionsBuilder;
+            _lobby = lobby;
+            _player = player;
```
DTO не так используют

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        private async void HandleLobbyHeartbeat()
+        {
+            if (_lobby.CurrentLobby == null) return;
+            if (_lobby.HostId == _player.Id) return;
```
А кто тогда будет биты посылать, если не хост?

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+            _heartbeatTimer -= Time.fixedDeltaTime;
+            if (_heartbeatTimer <= 0f)
+            {
+                float heartbeatTimerMax = 15;
```
вынести в константы

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        private void DeleteLobbyData()
+        {
+            _lobby = new Lobby();
+            _player.CurrentLobby = null;
```
зачем ты одну информацию в двух местах хранишь?

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        
+        #region CreateLobby
+        
+        private CreateLobbyOptions LobbyOptions(bool isPrivate, Unity.Services.Lobbies.Models.Player player, Dictionary<string, DataObject> data)
```
что LobbyOptions? Create, Delete, Update, Read?

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        
+        private CreateLobbyOptions LobbyOptions(bool isPrivate, Unity.Services.Lobbies.Models.Player player, Dictionary<string, DataObject> data)
+        {
+            _lobbyOptionsBuilder = new LobbyOptionsBuilder().Reset();
```
он в уонструкторе ресетается

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            _lobbyOptionsBuilder = new LobbyOptionsBuilder().Reset();
+            
+            _lobbyOptionsBuilder.SetIsPrivate(isPrivate).SetPlayer(player).SetData(data);
```
каждая команда в FluentAPI должна начинаться с новой строки

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            try
+            {
+                Unity.Services.Lobbies.Models.Lobby lobby = await Unity.Services.Lobbies.LobbyService.Instance.CreateLobbyAsync(lobbyName, maxPlayers, LobbyOptions(isPrivate, player.PlayerModel, data));
```
потом сто лет будешь дебажить вызов метода внутри вызова метода

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            try
+            {
+                await Lobbies.Instance.DeleteLobbyAsync(_lobby.CurrentLobby.Id);
```
нет проверки на роль игрока в лобби

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,128 @@
+            {
+                await Lobbies.Instance.RemovePlayerAsync(_lobby.CurrentLobby.Id, id);
+                Debug.Log($"Exit from {_lobby.CurrentLobby.Name}");
+                DeleteLobbyData();
```
try/Catch сильно просаживает производительность, не надо пихать в него что попало

Hitsar replied:
я вобще не вижу смысла в нём

Aher0nt commented on Assets/Scripts/UI/LobbyService/DeleteLobbyButton.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class DeleteLobbyButton : MonoBehaviour
+    {
+        [SerializeField] private Button _deleteLobbyButton;
```
мы кнопку сами вытаскиваем из TryGetCovponent

Hitsar replied:
а я говорил алексу добавить это >:|

Aher0nt commented on Assets/Scripts/UI/LobbyService/DeleteLobbyButton.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+            _lobbyService = lobbyService;
+
+            _deleteLobbyButton.onClick.AddListener(_lobbyService.DeleteLobby);
```
Этот код выполняется +- до загрузки сцены. Понятно, что это скорее всего работает, но методам юнити мы доверяем больше, пожтому работа с движком должна быть в методах движка, а инит только для протягивания зависимостей

Aher0nt commented on Assets/Scripts/UI/LobbyService/ExitLobbyButton.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class ExitLobbyButton : MonoBehaviour
+    {
+        [SerializeField] private Button _exitLobbyButton;
```
Через TryGetComponent

Aher0nt commented on Assets/Scripts/UI/LobbyService/ExitLobbyButton.cs:
```diff
@@ -0,0 +1,16 @@
+        [SerializeField] private Button _exitLobbyButton;
+        
+        [Inject]
+        private void Init(ILobbyService lobbyService, Player player) => _exitLobbyButton.onClick.AddListener(() => lobbyService.ExitLobby(player.Id));
```
В Awake/Start

Aher0nt replied:
Если так сильно нужен player, то надо стучаться с сервис аутентификации + Player к этому моменту еще пустой и вообще нельзя его в DI запихивать

Aher0nt commented on Assets/Scripts/UI/LobbyService/JoinLobbyButton.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class JoinLobbyButton : MonoBehaviour
+    {
+        [SerializeField] private TMP_InputField _lobbyCodeInputField;
```
Field можно не писать

Aher0nt commented on Assets/Scripts/UI/LobbyService/JoinLobbyButton.cs:
```diff
@@ -0,0 +1,17 @@
+        [SerializeField] private Button _joinLobbyButton;
+        
+        [Inject]
+        private void Init(ILobbyService lobbyService) => _joinLobbyButton.onClick.AddListener(() => lobbyService.JoinLobbyByCode(_lobbyCodeInputField.text));
```
в Awake/Start

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,45 @@
+    public class LobbiesDisplay : MonoBehaviour
+    {
+        [SerializeField] private Button _refreshButton;
+        private QueryLobbiesOptionsBuilder _queryLobbiesOptions;
```
В UI не должно быть логики

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace Assets.Scripts.UI.LobbyService
+{
+    public class LobbiesDisplay : MonoBehaviour
```
это навешивается на Button. Добавить в название класса

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    public class LobbiesDisplay : MonoBehaviour
+    {
+        [SerializeField] private Button _refreshButton;
```
через TryGetComponent

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,45 @@
+            {
+                Debug.Log(e);
+            }
+        }
```
В UI не должно быть логики

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,45 @@
+        {
+            _queryLobbiesOptions = new QueryLobbiesOptionsBuilder().SetCount(5);
+            return _queryLobbiesOptions.Build();
+        }
```
не должно

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbyCreatorUi.cs:
```diff
@@ -0,0 +1,23 @@
+        
+        [Inject]
+        private void Init(ILobbyService lobbyService) =>
+            _createNewLobbyButton.onClick.AddListener(() => lobbyService.CreateLobby(
```
В Awake/Start

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbyPlayersDisplay.cs:
```diff
@@ -0,0 +1,26 @@
+
+        public void PrintPlayers()
+        {
+            Lobby currentLobby = _player.CurrentLobby;
```
а нафиг тебе player, если ты с лобби работаешь?

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbyPlayersDisplay.cs:
```diff
@@ -0,0 +1,26 @@
+            Debug.Log("Players in Lobby " + currentLobby.Name);
+            foreach (Player player in currentLobby.Players)
+            {
+                Debug.Log("Player ID: " + player.Id + ", Player name: " + player.Name);
```
Надо создать инфраструктуру, чтобы это выводил другой компонент. Это не может вывести кнопка

Hitsar committed:
Реквест

надо исправить DTO

Aher0nt commented on Assets/Scripts/LobbyService/Builders/LobbyOptionsBuilder.cs:
```diff
@@ -0,0 +1,39 @@
+    {
+        private CreateLobbyOptions _lobbyOptions;
+
+        public LobbyOptionsBuilder SetIsPrivate(bool? isPrivate)
```
теперь у тебя при создании билдер невалидный.
вот такое
new LobbyOptionBuilder().Build() 
тоже должно работать

Aher0nt commented on Assets/Scripts/LobbyService/Builders/QueryLobbiesOptionsBuilder.cs:
```diff
@@ -0,0 +1,57 @@
+{
+    public class QueryLobbiesOptionsBuilder
+    {
+        private QueryLobbiesOptions _queryLobbiesOptions;
```
то же самое

Aher0nt commented on Assets/Scripts/LobbyService/Displays/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Assets.Scripts.LobbyService.Displays
+{
+    public class LobbiesDisplay : MonoBehaviour, ILobbiesDisplay
```
то, что ты перенес UI в другую папку, ничего не меняет

Aher0nt commented on Assets/Scripts/LobbyService/Displays/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,38 @@
+                Debug.Log("Lobbies found: " + queryResponse.Results.Count);
+                foreach (Unity.Services.Lobbies.Models.Lobby lobby in queryResponse.Results)
+                {
+                    Debug.Log($"{lobby.Name}, max players {lobby.MaxPlayers}");
```
это все еще нарушение SRP

Aher0nt commented on Assets/Scripts/LobbyService/Displays/LobbyPlayersDisplay.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Assets.Scripts.LobbyService.Displays
+{
+    public class LobbyPlayersDisplay : MonoBehaviour
```
это тоже UI

Aher0nt commented on Assets/Scripts/LobbyService/Displays/LobbyPlayersDisplay.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class LobbyPlayersDisplay : MonoBehaviour
+    {
+        private Lobby _lobby;
```
не используется

Hitsar replied:
используется, но не присваевается

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,117 @@
+        private Player _player;
+        private float _heartbeatTimer;
+
+        private const int HeartbeatTimerMax = 15;
```
константы пишутся капсом через нижнее подчеркивание

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,117 @@
+            LobbyOptionsBuilder lobbyOptionsBuilder = new LobbyOptionsBuilder().Reset();
+            
+            lobbyOptionsBuilder.SetIsPrivate(isPrivate)
+                .SetData(data);
```
а в чем смысл эти две операции в отдельную строчку выносить?

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,117 @@
+            _lobby.Name = lobby.Name;
+            _lobby.Id = lobby.Id;
+            _lobby.HostId = lobby.HostId;
+            _lobby.Players.Add(_player);
```
мы стараемся писать чистые функции
https://ru.wikipedia.org/wiki/Чистота_функции
а не вот это вот

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,117 @@
+            
+            Debug.Log($"Exit from {_lobby.Name}");
+            _lobby = new Lobby();
+            OnPlayerCountChange?.Invoke();
```
а теперь задачка посложнее. UI будет обновляться только у человека, который присоединился. Надо придумать механизм доставки события до других клиентов

Hitsar replied:
я в курсе, надо глянуть как вызывать метод у других клиентов

Aher0nt replied:
никак, тут нет сетевого программирования

Hitsar replied:
об этом я тоже в курсе, ну вот когда появится и сделаю, а пока просто гляну

Aher0nt replied:
так не появится...

Aher0nt replied:
так не появится...

Aher0nt commented on Assets/Scripts/UI/LobbyService/DeleteLobbyButton.cs:
```diff
@@ -0,0 +1,26 @@
+        
+        private void Start()
+        {
+            _deleteLobbyButton = GetComponent<Button>();
```
не GetComponent, а TryGetComponent. 
GetComponent обладает неопределенным поведением, мы им больше не пользуемся

Aher0nt commented on Assets/Scripts/UI/LobbyService/ExitLobbyButton.cs:
```diff
@@ -0,0 +1,24 @@
+        
+        private void Start()
+        {
+            _exitLobbyButton = GetComponent<Button>();
```
TryGetComponent

Aher0nt commented on Assets/Scripts/UI/LobbyService/JoinLobby.cs:
```diff
@@ -0,0 +1,26 @@
+        private void Start()
+        {
+            _joinLobbyButton = GetComponentInChildren<Button>();
+            _lobbyCodeInput = GetComponentInChildren<TMP_InputField>();
```
Так мы тоже не делаем. Никто не гарантирует, что у него в детях будут те самые компоненты, что они будут в единственном числе и прочее

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbiesFilter.cs:
```diff
@@ -0,0 +1,29 @@
+        private void Start()
+        {
+            _saveButton = GetComponentInChildren<Button>();
+            _countInputField = GetComponentInChildren<TMP_InputField>();
```
нет

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbyCreator.cs:
```diff
@@ -0,0 +1,33 @@
+        private void Start()
+        {
+            _createLobbyButton = GetComponentInChildren<Button>();
+            _isPrivateToggle = GetComponentInChildren<Toggle>();
```
Нет

Aher0nt commented on Assets/Scripts/UI/LobbyService/RefreshButton.cs:
```diff
@@ -0,0 +1,23 @@
+
+        private void Start()
+        {
+            _refreshButton = GetComponentInChildren<Button>();
```
его надо прямо на кнопку вешать, а не сбоку где-то

Hitsar replied:
это случайно написал

Hitsar committed:
рефакторинг часть 1

отделил логику от визуала, заменил GetComponent на Try, вернул конструктор билдерам, lobby dto создаётся и раздаётся эвентом

осталось разобраться с player DTO и клиентами

Aher0nt commented on Assets/Scripts/LobbyService/ILobbiesList.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface ILobbiesList
+    {
+        public event Action<string, int> OnGetLobbies;
```
название нарушает кодстайл, в дс я все писал

Aher0nt commented on Assets/Scripts/LobbyService/ILobbiesList.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        public event Action<string, int> OnGetLobbies;
+        
+        void GetLobbies();
```
это Get метод. он не может возвращать void

Hitsar replied:
Ты говорил что у нас такого не будет

Aher0nt commented on Assets/Scripts/LobbyService/ILobbiesList.cs:
```diff
@@ -0,0 +1,12 @@
+        public event Action<string, int> OnGetLobbies;
+        
+        void GetLobbies();
+        void FilterOptions(int count);
```
что такое "отфильтровать настройки" это как? Придумай методу нормальное название

Aher0nt commented on Assets/Scripts/LobbyService/ILobbyService.cs:
```diff
@@ -0,0 +1,17 @@
+    public interface ILobbyService
+    {
+        public event Action<Lobby> OnCreateLobby;
+        public event Action OnPlayerCountChange;
```
нарушает кодстайл

Aher0nt commented on Assets/Scripts/LobbyService/ILobbyService.cs:
```diff
@@ -0,0 +1,17 @@
+        public event Action<Lobby> OnCreateLobby;
+        public event Action OnPlayerCountChange;
+
+        void CreateLobby(string lobbyName, int maxPlayers, bool isPrivate, Dictionary<string, DataObject> data);
```
почему зависимости от сервисов юнити снова в интерфейсе? Убрать

Aher0nt commented on Assets/Scripts/LobbyService/LobbiesList.cs:
```diff
@@ -0,0 +1,36 @@
+            
+                Debug.Log("Lobbies found: " + queryResponse.Results.Count); 
+                foreach (Unity.Services.Lobbies.Models.Lobby lobby in queryResponse.Results)
+                    OnGetLobbies?.Invoke(lobby.Name, lobby.MaxPlayers);
```
ни в коем случае! Мы ответ от функции получаем только через return

Hitsar replied:
Мне сделать эвент который будет который будет вызывать метод у этого класса для  получения данных?

Aher0nt commented on Assets/Scripts/LobbyService/LobbyPlayersList.cs:
```diff
@@ -0,0 +1,32 @@
+            foreach (Player player in _lobby.Players)
+            {
+                
+                OnGetPlayers?.Invoke(player.Id, player.Name);
```
нет

Aher0nt commented on Assets/Scripts/LobbyService/LobbyService.cs:
```diff
@@ -0,0 +1,125 @@
+
+        public async void ExitLobby(string id)
+        {
+            if (_lobby.HostId == null) return;
```
Так у тебя в таком случае NullPointerException просто будет

Hitsar replied:
тогда эта проверка не нужна?

Aher0nt commented on Assets/Scripts/UI/LobbyService/DeleteLobbyButton.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Assets.Scripts.UI.LobbyService
+{
+    public class DeleteLobbyButton : MonoBehaviour
```
RequireComponent добавь

Aher0nt commented on Assets/Scripts/UI/LobbyService/Displays/LobbiesDisplay.cs:
```diff
@@ -0,0 +1,21 @@
+
+        private void Start() => _lobbiesList.OnGetLobbies += DisplayLobbies;
+        
+        private void DisplayLobbies(string name, int maxPlayer)
```
так ты много лобби показываешь, или одно?

Aher0nt commented on Assets/Scripts/UI/LobbyService/Displays/LobbyPlayersDisplay.cs:
```diff
@@ -0,0 +1,21 @@
+
+        private void Start() => _playersList.OnGetPlayers += DisplayPlayers;
+        
+        private void DisplayPlayers(string id, string name)
```
название не соответствует аргументам

Hitsar replied:
Название метода: отобразить игроков, думаю не сложно догадаться что id и name игрока
если ты про множ. число то уже поменял

Aher0nt commented on Assets/Scripts/UI/LobbyService/ExitLobbyButton.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        private Button _exitLobbyButton;
+        private ILobbyService _lobbyService;
+        private Player _player;
```
UI не знает про модель игрока

Aher0nt commented on Assets/Scripts/UI/LobbyService/JoinLobby.cs:
```diff
@@ -0,0 +1,26 @@
+        private void Start()
+        {
+            TryGetComponent(out _joinLobbyButton);           
+            TryGetComponent(out _lobbyCodeInput); 
```
и как ты себе это представляешь? кнопку с полем ввода

Hitsar replied:
Ну так скрипт не для кнопки

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbyCreator.cs:
```diff
@@ -0,0 +1,33 @@
+        [SerializeField] private TMP_InputField _maxPlayersInputField;
+        
+        private Button _createLobbyButton;
+        private Toggle _isPrivateToggle;
```
Все компоненты надо получать одним способом

Hitsar replied:
Как мне 2 одинаковых компонента взять, и быть уверенным что они возьмутся правильно

Aher0nt commented on Assets/Scripts/UI/LobbyService/LobbyCreator.cs:
```diff
@@ -0,0 +1,33 @@
+            
+            _createLobbyButton.onClick.AddListener(() => _lobbyService.CreateLobby(
+                _lobbyNameInputField.text,
+                int.Parse(_maxPlayersInputField.text),
```
Если ввести строку, вылетит исключение. Исключения надо обрабатывать
